src/beamlime/applications/handlers.py

```diff
@@ -3,37 +3,59 @@
 import argparse
 import pathlib
 from dataclasses import dataclass
-from typing import Any, NewType
+from typing import NewType
 
 import scipp as sc
+from scippneutron.io.nexus._json_nexus import JSONGroup
 
 from beamlime.logging import BeamlimeLogger
 
 from ..stateless_workflow import StatelessWorkflow, WorkflowResult
-from .base import HandlerInterface, MessageProtocol
+from .base import HandlerInterface
+from .daemons import DetectorDataReceived, RunStart
 
 Events = NewType("Events", list[sc.DataArray])
 
 
+@dataclass
+class DataReady:
+    content: JSONGroup
+
+
 @dataclass
 class WorkflowResultUpdate:
     content: WorkflowResult
 
 
+class DataAssembler(HandlerInterface):
+    """Receives data and assembles it into a single data structure."""
+
+    def __init__(self) -> None:
+        from ._nexus_helpers import NexusContainer
+
+        self.nexus_container: NexusContainer
+
+    def set_run_start(self, message: RunStart) -> None:
+        self.nexus_container = message.content
+
+    def assemble_detector_data(self, message: DetectorDataReceived) -> DataReady:
+        self.info("Received data from detector %s", message.content['source_name'])
+        self.nexus_container.insert_ev44(message.content)
+        return DataReady(
+            content=self.nexus_container.nexus_dict  # should be a JSONGroup later.
+        )
```

Comment:
This looks like it is inserting only a single detector, so the data passed to the data reduction always only contains a single (but different every time) detector? I think this will not work. The listener must merge this data (as mentioned in the issue #134), such that the assembled result is roughly equivalent to reading a time-slice from the file.

Reply:
No, the ``insert_ev44`` method also replaces the ``nexus_dict`` with the updated one, so it still has other detectors.

Reply:
So this just accumulates indefinitely? I thought we wanted to accumulate for some time and *not* push the entire data to the data reduction with every update.

Reply:
No, it replaces the dataset for each detector.
We only accumulate the dataset of the number of detectors(ev44 place holders) at maximum.

Merging messages is not implemented yet.

Reply:
Should we have a call instead...?

Reply:
Ok, so you are saying that:
- Every time *one* detector gets an update...
- ... we replace that but keep the others
- Which means we will always have the "most recent" state (except at the beginning).
- We forward to data reduction *with every update*

Correct?

I think this will still not work, since at the beginning not all detectors will have received updates yet. Also, we call data reduction to often.

I am not saying that message merging needs implementing here, but shouldn't we gather updates from all detectors, such that what we pass to data reduction is consistent?

Reply:
> Correct?

Yes, the idea was to keep the ``most recent`` state.

> I am not saying that message merging needs implementing here, but shouldn't we gather updates from all detectors, such that what we pass to data reduction is consistent?

I thought that should be done as a part of ``merge``.
It's not so simple when you don't know when exactly all the detectors are sending their messages.
So we have to either
1. Flush the dataset(Trigger the data reduction) only when whole set of dataset from all detectors arrive
2. Merge `n*detectors` number of messages, hoping that they are usually received in order.

And I thought the option `1.` is not really feasible or robust because then we know which detector are sending data or not from the beginning. So it'll be broken as soon as part of detectors are not functioning or turned off for some reason. And even if we go for this option, aligning multiple messages based on the `source` should be done as a separate task not here.


Is this current implementation too far to be considered 'working'...?

Reply:
I think it will be

3. Merge M seconds, use empty data for datasets without updates.

But yes, we can do it in the follow up. I will write this down more clearly in an issue.

Reply:
I updated https://github.com/scipp/beamlime/issues/133. Please read the description of that task an let me know what is unclear!